### PR TITLE
Enable calls to GenAI-Perf for profile subcommand

### PR DESCRIPTION
**Pull request by dyastremsky**
This pull request makes it so that Triton CLI calls [GenAI-Perf](https://github.com/triton-inference-server/client/tree/main/src/c%2B%2B/perf_analyzer/genai-perf) for its profile subcommand. All arguments get passed through to GenAI-Perf.

As part of these changes, the previous profiler functionality has been removed from Triton CLI to avoid maintaining this behavior in both places.

Unit tests have successfully passed with these changes in place in an environment with PA and GenAI-Perf.
image

GenAI-Perf demo below. _Apologies for the extra errors beforehand from some arg types. Note: This was a previous iteration, `--task-llm` is no longer used. The reason the output tokens are off is because the mock Python model doesn't actually use the max_token inputs, it just returns the input as output._
https://github.com/triton-inference-server/triton_cli/assets/58150256/97056e68-afb8-49e9-9e07-8f6601952c3a

### CODE REVIEW

File: tests/test_models/mock_llm/config.pbtxt
Comment (dyastremsky) [inline tests/test_models/mock_llm/config.pbtxt:32]: The input names are being made lowercase to match what GenAi-Perf generates for TRT-LLM/vLLM. Max_tokens is being added as an arg so that this doesn't break in the future. There is an [upcoming PR](https://github.com/triton-inference-server/client/pull/612) that will make sure that PA returns an error if a user-supplied input does not exist in the model (current behavior is to ignore it).

File: .gitignore
Comment (fpetrini15) [inline .gitignore:22]: Was this a leftover change or intentional? I believe we should not be ignoring the `tests` folder, right?

Comment (dyastremsky) [reply .gitignore]: Thanks, good catch! I was removing some artifacts and then did the parent-level directory. I forgot these have actual tests in them. Just pushed a commit to be more specific with these exclusions.

File: src/triton_cli/parser.py
Comment (fpetrini15) [inline src/triton_cli/parser.py:512]:
```suggestion
            cmd = cmd + [f"-{arg}", f"{value}"] if len(arg)==1 else cmd + [f"--{arg}", f"{value}"]
```
Maybe same thing here? If not, I think we can get rid of the `arg=arg` statement.

File: src/triton_cli/parser.py
Comment (fpetrini15) [inline src/triton_cli/parser.py:522]: What scenario are we guarding against here? In my testing no args are ever pruned and argv is always something like this:
`['profile', '-m', 'gpt2', '--task', 'llm', '--concurrency', '1']`

**Comment by fpetrini15** [comment]
Great work David! Only some minor tweaks and clarifications.

Just a thought, should we be setting a default value for the required arguments? For example, the `profile` subcommand requires `concurrency` as a parameter. During the arg sanitation process, would it make sense to add a default argument for it if the user didn't provide it. I think this would abstract some complexity away from the user.

CC: @rmccorm4 for thoughts.

File: src/triton_cli/parser.py
Comment (dyastremsky) [reply src/triton_cli/parser.py:506]: Is this how we do this in the rest of our Python code? I generally like if/else statements because it's clear what's happening (the other code is more concise, but slower and less clear to read IMO). I'm happy to follow whatever convention we generally use though for consistency.

File: src/triton_cli/parser.py
Comment (dyastremsky) [reply src/triton_cli/parser.py:484]: This one should be fine, because if argv is None, it will read from the command line. This preserves the regular behavior for all commands that do not require pass-through behavior.

File: src/triton_cli/parser.py
Comment (dyastremsky) [reply src/triton_cli/parser.py]: That's an embarrassing oversight. Thanks for catching this! Removed.

Comment (dyastremsky) [reply src/triton_cli/parser.py]: Left a comment above regarding the conditional style.

File: src/triton_cli/parser.py
Comment (dyastremsky) [reply src/triton_cli/parser.py:521]: Updated this comment to clarify. Try running `triton optimize profile <args>` (like `triton optimize profile --model-repository models \
    --profile-models add_sub --triton-launch-mode=docker \
    --output-model-repository-path output_dir \
    --export-path profile_results`). Without this logic, it will fail, since the parser thinks the profile is another subcommand.

**Comment by dyastremsky** [comment]
> Great work David! Only some minor tweaks and clarifications.
> 
> Just a thought, should we be setting a default value for the required arguments? For example, the `profile` subcommand requires `concurrency` as a parameter. During the arg sanitation process, would it make sense to add a default argument for it if the user didn't provide it. I think this would abstract some complexity away from the user.
> 
> CC: @rmccorm4 for thoughts.

I think we're 100% going for a passthrough approach here. This makes the Triton CLI extensible and maintains all logic unique to the tools in their own repositories. If we start moving over logic, then responsibilities are starting to overlap and we could be duplicating code. If there is a required arg for profile and we think it should set a default, Model Analyzer is the right place to fix that.

File: src/triton_cli/parser.py
Comment (fpetrini15) [reply src/triton_cli/parser.py:521]: That's the use-case I was looking for, thanks David!

File: src/triton_cli/parser.py
Comment (fpetrini15) [reply src/triton_cli/parser.py]: Not embarrassing; that's why we have reviews! :)

File: src/triton_cli/parser.py
Comment (fpetrini15) [reply src/triton_cli/parser.py:506]: It's definitely an optional and stylistic choice. I, personally, just find it more readable when the logic within each block is a a initialization-like one-liner. But I'm happy to leave it as-is as well!

File: src/triton_cli/parser.py
Comment (rmccorm4) [reply src/triton_cli/parser.py:521]: `triton profile` and `triton optimize profile` :thinking: 

Naming things is hard..

File: src/triton_cli/constants.py
Comment (rmccorm4) [inline src/triton_cli/constants.py:49]: Moving to thread :thread:

Comment (rmccorm4) [reply src/triton_cli/constants.py:49]: @dyastremsky :

> I think we're 100% going for a passthrough approach here. This makes the Triton CLI extensible and maintains all logic unique to the tools in their own repositories. If we start moving over logic, then responsibilities are starting to overlap and we could be duplicating code. If there is a required arg for profile and we think it should set a default, Model Analyzer is the right place to fix that.

File: src/triton_cli/parser.py
Comment (dyastremsky) [reply src/triton_cli/parser.py:506]: I'm super open. I want to be consistent with the rest of our codebase. If we don't have a general rule on this yet, I'll leave it as is if that's okay. I'm finding documentation online that the ternary operator is best reserved for simple cases, and this is less simple for me then something like `min = a if a < b else b`.

File: src/triton_cli/constants.py
Comment (rmccorm4) [reply src/triton_cli/constants.py:49]: Agreed that passthrough as much as possible will be the simplest to maintain, and fixing ergonomics in the downstream tools being invoked would be preferred. There may be some cases to fork logic in CLI, but should be few and far between if we can help it.

File: src/triton_cli/parser.py
Comment (fpetrini15) [reply src/triton_cli/parser.py:506]: I don't believe we have a general rule for this project, so let's just leave as-is.

File: tests/test_e2e.py
Comment (rmccorm4) [inline tests/test_e2e.py:70]: I think this is failing the `pytest` runs in the github-side checks because there is no `model-analyzer` command found. Can you add the installation as a dependency in pyproject.toml and any other test fix necessary to see if we can get the :heavy_check_mark: on github side?

Comment (rmccorm4) [reply tests/test_e2e.py:70]: ex: https://github.com/triton-inference-server/triton_cli/actions/runs/8914215110/job/24481334037?pr=52

```
test_e2e.py:82: in _optimize
...
args = ['model-analyzer', 'profile', '--profile-models', 'add_sub', '--triton-launch-mode', 'local', ...]
...
E               FileNotFoundError: [Errno 2] No such file or directory: 'model-analyzer'
```

Comment (dyastremsky) [reply tests/test_e2e.py:70]: Thanks for the suggestion, already done. :) I added it on the unit testing and documentation [PR (draft)](https://github.com/triton-inference-server/triton_cli/pull/53) that I plan to merge into this one once ready and approved. But I can move that work here so that this is passing even without those PRs.

I assume I'll need to clone and install GenAI-Perf more manually, since it doesn't have a pip wheel. Apologies that the these changes are not in place yet. I'm learning a lot about the triton_cli and the tool packaging this week. Originally, I thought this would be a much harder blocker than it is (i.e. that these tests might run in a Triton container or that I would need to get perf_analyzer installed on the GitHub action side). I can work through this.

**Comment by dyastremsky** [comment]
This work is currently on hold. I will comment the ticket once that status changes.

**Comment by dyastremsky** [comment]
@rmccorm4 @fpetrini15 This should now be ready, but it is failing the Python 3.8 test due to types not yet been subscriptable until 3.9. While I could go into GenAI-Perf and make it Python 3.8-friendly, given that's not a requirement of GenAI-Perf, something will likely break again in the future. If we want this to pin the 24.04 version, this will also delay release until there is a release branch with the changes in GenAI-Perf.

**Comment by dyastremsky** [comment]
> @rmccorm4 @fpetrini15 This should now be ready, but it is failing the Python 3.8 test due to types not yet been subscriptable until 3.9. While I could go into GenAI-Perf and make it Python 3.8-friendly, given that's not a requirement of GenAI-Perf, something will likely break again in the future. If we want this to pin the 24.04 version, this will also delay release until there is a release branch with the changes in GenAI-Perf.

I tried creating a branch with types that work with Python 3.8 ([here](https://github.com/triton-inference-server/client/tree/dyas-types)), but then it's running into issues with not having the built-in library itertools available in 3.10. I'd need to go back and retrofit code. Is Python 3.8 support a requirement for Triton CLI? It seems for this integration, we may need to either make it a requirement for GenAI-Perf or not a requirement for either.

CC: @debermudez @nnshah1

**Comment by rmccorm4** [comment]
It's not a strong requirement at this time. I added it to CLI because it was easy and I believe PyTriton supported 3.8+

I don't mind removing the support for now if it's not a simple fix to support or just unwanted.

**Comment by dyastremsky** [comment]
> I don't think it's a strong requirement at this time. I added it to CLI because it was easy and I believe PyTriton supported 3.8+
> 
> I don't mind removing the support for now if it's not a simple fix to support or just unwanted.

If you're okay with it, I pushed a commit dropping it. I understand the desire for greater support though, so if we wanted to support 3.8-3.9, it would just require updating GenAI-Perf to remove the 3.10+ features and then test on 3.8. I'll start a conversation about it in the morning.

File: README.md
Comment (rmccorm4) [inline README.md:100]:
```suggestion
```

File: README.md
Comment (rmccorm4) [inline README.md:229]: Has this not been updated in genai-perf yet? I thought genai-perf was already updated to use `tensorrtllm` instead

File: src/triton_cli/parser.py
Comment (rmccorm4) [inline src/triton_cli/parser.py:453]: Can you move all these helpers to `profile.py` and then:
```python
from profile import build_command
...
def handle_profile(args: argparse.Namespace):
    cmd = profile.build_command(args, executable="genai-perf")
    logger.info(f"Running: '{' '.join(cmd)}'")
    subprocess.run(cmd, check=True)
```

File: README.md
Comment (rmccorm4) [reply README.md:227]: Oh I see in your other comments it's coming in 24.05, hmm. I am against changing this flag back and forth or being consistent for the next few weeks.

Can the special arg parse handling you're already doing for the `profile` subcommand just add a special exception to map `--backend tensorrtllm` to `--backend trtllm` until fixed in later genai-perf version?

Don't want to change the front-facing args on Triton CLI

File: src/triton_cli/parser.py
Comment (dyastremsky) [reply src/triton_cli/parser.py]: Sure, done!

File: README.md
Comment (dyastremsky) [reply README.md]: It will be updated in 24.05, but is not updated in 24.04. I updated the logic to do a direct substitution of tensorrtllm to trtllm when passed to GenAI-Perf but missed updating it here. I did so now. I have an open ticket to fix that substitution hotfix once 24.05 is released and used in the CLI.

File: README.md
Comment (rmccorm4) [reply README.md:100]: Did this get added right back? 🤔

File: pyproject.toml
Comment (rmccorm4) [inline pyproject.toml:52]: Can you update the `tritonclient==<version>` dependency a few lines down to match 24.04 version so we get the right PA version too?

File: src/triton_cli/parser.py
Comment (rmccorm4) [inline src/triton_cli/parser.py:369]: Can you show me the `triton -h` and `triton profile -h` outputs after these changes?

File: tests/test_cli.py
Comment (rmccorm4) [inline tests/test_cli.py:139]: Can you ELI5 or add a comment on what this test is doing? Is it just asserting that the command gets transformed to the right genai-perf command correctly?

File: README.md
Comment (dyastremsky) [reply README.md]: Sorry, my page didn't upload to reflect your latest comment. Yes, that's exactly what's being done. :) Just didn't reflect it here.

I dropped the special arg handling though, since we're not going to be handling that case anymore.

File: tests/test_e2e.py
Comment (rmccorm4) [inline tests/test_e2e.py:160]: thanks for the fix here 🚀

**Review by rmccorm4 — COMMENTED** [review]
Looking really good! Only minor comments

File: src/triton_cli/parser.py
Comment (dyastremsky) [reply src/triton_cli/parser.py:369]: Good question! Sure.

`triton -h`:
image

`triton profile -h` (cuts off at the end, this is the GenAI-Perf help output):
image

File: pyproject.toml
Comment (dyastremsky) [reply pyproject.toml:52]: Good call. Updated.

File: tests/test_cli.py
Comment (dyastremsky) [reply tests/test_cli.py:141]: Yes, exactly. :) It mocks the subprocess call to make sure the right subprocess call gets made once (but not actually executed). I added a comment.

File: src/triton_cli/profile.py
Comment (rmccorm4) [inline src/triton_cli/profile.py:53]: Now that I see the output of `triton profile -h` showing genai-perf help output mentioning `--backend trtllm`, I think we should support both just for intuition for those reading the help output.


```suggestion
            if value in ["tensorrtllm", "trtllm"]:
```

File: src/triton_cli/profile.py
Comment (rmccorm4) [inline src/triton_cli/profile.py:54]: I think we are skipping passing through the arg when `--backend` is passed with a different value like `vllm`. We should passthrough any other value and let genai-perf raise the error if it's an unsupported backend

**Comment by dyastremsky** [comment]
As per the offline discussion, I have removed the 3.8 test for now.

@rmccorm4 @fpetrini15 This is ready for another round of review.

File: tests/test_e2e.py
Comment (rmccorm4) [inline tests/test_e2e.py:152]: Why is this test removed? Can we keep it in for the `infer()` part even if we don't do `profile()`? Want to keep some sanity test that `triton infer` is working on arbitrary models.

File: tests/test_e2e.py
Comment (rmccorm4) [inline tests/test_e2e.py:114]: Just a question for my own future knowledge, these changes don't need to be added in this PR:
- is there anything interesting about genai-perf we can also test for here? 
- ex: parsing an output CSV file for expected values or something? 
- ex: testing both streaming and non-streaming
- ...

File: src/triton_cli/parser.py
Comment (rmccorm4) [inline src/triton_cli/parser.py:442]:
```suggestion
    # Add special argparse handling for passthrough to genai-perf CLI
    if argv_[0] == "profile":
```

File: src/triton_cli/__init__.py
Comment (rmccorm4) [inline src/triton_cli/__init__.py:27]: I want it to be clear from terminal outputs, screenshots, copy/paste, etc. that this is installed from source and not a release to ease future debugging/issues
```suggestion
__version__ = "0.0.8dev"
```

Comment (rmccorm4) [reply src/triton_cli/__init__.py:27]: Similar to here: https://github.com/triton-inference-server/server/blob/main/TRITON_VERSION

File: tests/test_e2e.py
Comment (dyastremsky) [reply tests/test_e2e.py:114]: We could look to ensure there is some form of output (e.g. the expected chart name and/or some of the metrics). The tool is still in the alpha phase with a lot of changes, so those tests here could end up breaking as things change there.

I think the zero exit code should be sufficient to know that it ran successfully though.

File: tests/test_e2e.py
Comment (dyastremsky) [reply tests/test_e2e.py:152]: Sure, let me return this. This was based on our conversations about Triton CLI not officially supporting non-LLM inference at the moment. I thought I added this test and missed that it's being removed in my review of the PR.

File: tests/test_e2e.py
Comment (rmccorm4) [inline tests/test_e2e.py:166]: Need to remove this whole `with pytest.raises(Exception)` block: https://github.com/triton-inference-server/triton_cli/blob/07bd3379d709d66a1df97750a5dc40bd69f11c18/tests/test_e2e.py#L158-L161

**Review by rmccorm4 — COMMENTED** [review]
LGTM other than this fix for `test_non_llm`: https://github.com/triton-inference-server/triton_cli/pull/52/files#r1596116006

Nice work David! :rocket:

File: tests/test_e2e.py
Comment (dyastremsky) [reply tests/test_e2e.py:166]: Ah, thanks. Removed!

**Review by rmccorm4 — APPROVED** [review]
:partying_face:

**Pull request merged**
